rclpy/src/rclpy/publisher.cpp

```diff
@@ -0,0 +1,205 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+
+#include <memory>
+#include <string>
+
+extern "C"
+{
+#include "rclpy_common/common.h"
+}
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+
+#include "publisher.hpp"
+
+namespace rclpy
+{
+static void
+_rclpy_destroy_publisher(void * p)
+{
+  auto pub = static_cast<rclpy_publisher_t *>(p);
+  if (!pub) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_rclpy_destroy_publisher got NULL pointer");
+    return;
+  }
+
+  rcl_ret_t ret = rcl_publisher_fini(&(pub->publisher), pub->node);
+  if (RCL_RET_OK != ret) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini publisher: %s",
+      rcl_get_error_string().str);
+  }
+  PyMem_Free(pub);
+}
+
+py::capsule
+publisher_create(
+  py::capsule pynode, py::object pymsg_type, std::string topic,
+  py::capsule pyqos_profile)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  auto msg_type = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!msg_type) {
+    throw py::error_already_set();
+  }
+
+  rcl_publisher_options_t publisher_ops = rcl_publisher_get_default_options();
+
+  if (!pyqos_profile.is_none()) {
+    if (0 != strcmp("rmw_qos_profile_t", pyqos_profile.name())) {
+      throw py::value_error("capsule is not an rmw_qos_profile_t");
+    }
+    auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
+    publisher_ops.qos = *qos_profile;
+  }
+
+  // Use smart pointer to make sure memory is free'd on error
+  auto deleter = [](rclpy_publisher_t * ptr) {_rclpy_destroy_publisher(ptr);};
+  auto pub = std::unique_ptr<rclpy_publisher_t, decltype(deleter)>(
+    static_cast<rclpy_publisher_t *>(PyMem_Malloc(sizeof(rclpy_publisher_t))),
+    deleter);
+  if (!pub) {
+    throw std::bad_alloc();
+  }
+  pub->publisher = rcl_get_zero_initialized_publisher();
+  pub->node = node;
+
+  rcl_ret_t ret = rcl_publisher_init(
+    &(pub->publisher), node, msg_type,
+    topic.c_str(), &publisher_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
+      std::string error_text{"Failed to create publisher due to invalid topic name '"};
+      error_text += topic;
+      error_text += "'";
+      throw RCLError(error_text);
```

Comment:
I think it should be `py::value_error` when the topic name is invalid to [match the documentation](https://github.com/ros2/rclpy/pull/695/files#diff-68392ec0fdc86b592da9d13cbf7f93f9ddeafb0b43b5070804574cbdf1b5f851R32).

Reply:
658b6f5